README.md
# Domoticz Plugin homematicIP Window and Door Contact with magnet (HMIP-SWDM)
Domoticz plugin for the homematicIP Window and Door Contact with magnet (HMIP-SWDM).

# Objectives
* To monitor the state open / close of windows and doors
* To alert via notification in case state is open using Domoticz dzVents script

![domoticz-plugin-hmip-swdm-o](https://user-images.githubusercontent.com/47274144/105820917-79fb4c00-5fba-11eb-8acc-da71fd4bb98a.png)

## Solution
To monitor the state open / close of windows and doors the homematicIP Window and Door Contact with magnet (HMIP-SWDM) is used.
The HMIP-SWDM is connected to a homematic IP system.
The homematic IP system used is a [RaspberryMatic](https://raspberrymatic.de/) operating system running the Homematic Central-Control-Unit (CCU).
The CCU has the additional software XML-API CCU Addon installed.
Communication between Domoticz and the CCU is via HTTP XML-API requests with HTTP XML response.

In Domoticz, following device(s) is/are created:
(Type,SubType) [XML-API Device Datapoint Type] - Note)
* Alert (General,Alert) [STATE] - closed (green) or open (red)

The device state is updated every 60 seconds (default).

If required, further actions can defined, by for example creating a dzVents script to send a notification / email (see below "dzVents Example").

## Hardware
Hardware subject to change.
* Raspberry Pi 3B+ (RaspberryMatic System)
* homematicIP Window and Door Contact with magnet (HMIP-SWDM)

## Software
Software versions subject to change.
* Raspberry Pi OS ( Raspbian GNU/Linux 10 buster, kernel 5.4.83-v7l+)
* Domoticz 2020.2 (build 12847)
* RaspberryMatic 3.55.5.20201226 [info](https://raspberrymatic.de/)
* XML-API CCU Addon 1.20 [info](https://github.com/jens-maus/XML-API)
* Python 3.7.3
* Python module ElementTree

**Note on the Python Module ElementTree**
The Python Module **ElementTree XML API** is used to parse the XML-API response.
This module is part of the standard package and provides limited support for XPath expressions for locating elements in a tree. 

_Hint_
(Optional)
For full XPath support install the module **ElementPath** from the terminal command-line for Python 2.x and 3.x via pip:
``` 
sudo pip install elementpath
sudo pip3 install elementpath
```

## RaspberryMatic Prepare
The RaspberryMatic system has been setup according [these](https://github.com/jens-maus/RaspberryMatic) guidelines.

The XML-API CCU Addon is required and installed via the HomeMatic WebUI > Settings > Control panel > Additional software (download the latest version from previous URL shared).
**IMPORTANT**
Be aware of the security risk, in case the HomeMatic Control Center can be reached via the Internet without special protection (see XML-API Guidelines).

### XML-API Scripts
The XML-API provides various tool scripts, i.e. devices state list, device state or set new value and many more.
The scripts are submitted via HTTP XML-API requests.
The plugin makes selective use of scripts with device id and datapoint id's.
The device id is required to get the state of the device datapoints. The datapoint id's are required to get the state/value of device attributes.

#### Device ID (statelist.cgi)
Get Device ID (attribute "ise_id") from list of all devices with channels and current values: http://ccu-ip-address/addons/xmlapi/statelist.cgi.

From the HTTP XML-API response, the Device ID ("ise_id") is selected by searching for the
* Device Name (i.e. Garagentor) or 
* Device Channel (HmIP-SWDM 001558A99D5A78:1). 
The data is obtained from the HomeMatic WebUI Home page > Status and control > Devices.
The Device "ise_id" is required for the plugin parameter _Mode1_.
HTTP XML-API response: Device ID = 3597.
```
...
<device name="Garagentor" ise_id="3597" unreach="false" config_pending="false">
	<channel...>
</device>
...
```
This script 
* has to be run once from a browser prior installing the plugin to get the device id as required by the plugin parameter Device ID ("mode1") and the next script.
* is not used in the plugin.

#### Channel Datapoint(s) (state.cgi)
Request the Channel Datapoint(s) for a Device ID to get value(s) for selected attribute(s): http://ccu-ip-address/addons/xmlapi/state.cgi?device_id=DEVICE_ISE_ID
The **Device ID 3597** is used as parameter to get the device state from which the attributes can be selected. 
The HTTP XML-API response lists three Channels from which **Channel HmIP-SWDM 001558A99D5A78:1** (as previous shown in the HomeMatic WebUI) is used.
The datapoint(s) used:
* type="STATE" with ise_id="3622" to get the state of the contact. The value is from valuetype 16, i.e. 1 (open) or 0 (closed).
The datapoint "ise_id" is required for the plugin parameter _Mode2_.
```
<device name="Garagentor" ise_id="3597" unreach="false" config_pending="false">
	<channel name="Garagentor:0" ise_id="3598" index="0" visible="true" operate="true">
		<datapoint name="HmIP-RF.001558A99D5A78:0.CONFIG_PENDING" type="CONFIG_PENDING" ise_id="3599" value="false" valuetype="2" valueunit="" timestamp="1611044101" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.DUTY_CYCLE" type="DUTY_CYCLE" ise_id="3603" value="false" valuetype="2" valueunit="" timestamp="1611044101" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.LOW_BAT" type="LOW_BAT" ise_id="3605" value="false" valuetype="2" valueunit="" timestamp="1611044101" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.OPERATING_VOLTAGE" type="OPERATING_VOLTAGE" ise_id="3609" value="2.800000" valuetype="4" valueunit="" timestamp="1611044101" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.OPERATING_VOLTAGE_STATUS" type="OPERATING_VOLTAGE_STATUS" ise_id="3610" value="0" valuetype="16" valueunit="" timestamp="1611044101" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.RSSI_DEVICE" type="RSSI_DEVICE" ise_id="3611" value="178" valuetype="8" valueunit="" timestamp="1611044101" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.RSSI_PEER" type="RSSI_PEER" ise_id="3612" value="0" valuetype="8" valueunit="" timestamp="0" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.UNREACH" type="UNREACH" ise_id="3613" value="false" valuetype="2" valueunit="" timestamp="1611044101" operations="5"/>
		<datapoint name="HmIP-RF.001558A99D5A78:0.UPDATE_PENDING" type="UPDATE_PENDING" ise_id="3617" value="false" valuetype="2" valueunit="" timestamp="1610968122" operations="5"/>
	</channel>
	<channel name="HmIP-SWDM 001558A99D5A78:1" ise_id="3621" index="1" visible="true" operate="true">
		<datapoint name="HmIP-RF.001558A99D5A78:1.STATE" type="STATE" ise_id="3622" value="0" valuetype="16" valueunit="""" timestamp="1611044101" operations="5"/>
	</channel>
	<channel name="HmIP-SWDM 001558A99D5A78:2" ise_id="3623" index="2" visible="true" operate="true"/>
</device>
```
This script 
* has to be run once from a browser prior installing the plugin to get the datapoint id as required by the plugin hardware Datapoint ID STATE ("mode2").
* is used in the plugin to get the device state in regular check intervals.
#### Change Value (statechange.cgi)
Change the State or Value for a Datapoint: http://ccu-ip-address/addons/xmlapi/statechange.cgi?ise_id=DATAPOINT_ISE_ID&new_value=NEW_VALUE
This script 
* is not used by the plugin.

#### Summary
The device id "3597" (for the device named "Garagentor") is used to 
* get the state "0" (closed") or "1" (open") of the channel "HmIP-SWDM 001558A99D5A78:1" datapoint type "STATE", ise_id "3622". 

## Domoticz Prepare
Open in a browser, four tabs with the Domoticz GUI Tabs: 
* Setup > Hardware = to add / delete the new hardware
* Setup > Devices = to check the devices created by the new hardware (use button Refresh to get the latest values)
* Setup > Log = to check the installation and check interval cycles for errors
* Active Menu depending Domoticz Devices created/used = to check the devices value
Ensure to have the latest Domoticz version installed: Domoticz GUI Tab Setup > Check for Update

### Domoticz Plugin Installation

### Plugin Folder and File
Each plugin requires a dedicated folder which contains the plugin, mandatory named **plugin.py**.
The folder is named according omematic IP device name. 
``` 
mkdir /home/pi/domoticz/plugins/hmip-swdm
``` 

Copy the file **plugin.py** to the folder.

### Restart Domoticz
``` 
sudo service domoticz.sh restart
``` 

### Domoticz Add Hardware
**IMPORTANT**
Prior adding the hardware, set in Domoticz GUI > Settings the option to allow new hardware.
If this option is not enabled, no new devices are created.
Check the GUI > Setup > Log as error message Python script at the line where the new device is used
(i.e. Domoticz.Debug("Device created: "+Devices[1].Name))

In the GUI > Setup > Hardware add the new hardware **homematicIP Window and Door Contact with Magnet (HmIP-SWDM)**.
Define the hardware parameter:
* CCU IP: The IP address of the homematic CCU. Default: 192.168.1.225.
* Device ID: The device datapoint ise_id - taken from the XMLAPI statelist request. Default: 3597.
* Datapoint STATE: The STATE datapoint ise_id - taken from the XMLAPI statelist request. Default: 3622.
* Check Interval (sec): How often the state of the device is checked. Default: 60.
* Debug: Set initially to true. If the plugin runs fine, update to false.

### Add Hardware - Check the Domoticz Log
After adding, ensure to check the Domoticz Log (GUI > Setup > Log)
Example - with hardware name C1 (just a name for testing instead using german name "Garagentor")
```
2021-01-25 13:26:15.675 (Garagentor) Debug logging mask set to: PYTHON PLUGIN QUEUE IMAGE DEVICE CONNECTION MESSAGE ALL
2021-01-25 13:26:15.675 (Garagentor) 'HardwareID':'14'
2021-01-25 13:26:15.675 (Garagentor) 'HomeFolder':'/home/pi/domoticz/plugins/hmip-swdm/'
2021-01-25 13:26:15.675 (Garagentor) 'StartupFolder':'/home/pi/domoticz/'
2021-01-25 13:26:15.675 (Garagentor) 'UserDataFolder':'/home/pi/domoticz/'
2021-01-25 13:26:15.675 (Garagentor) 'Database':'/home/pi/domoticz/domoticz.db'
2021-01-25 13:26:15.675 (Garagentor) 'Language':'en'
2021-01-25 13:26:15.675 (Garagentor) 'Version':'1.1.1 (Build 20210125)'
2021-01-25 13:26:15.675 (Garagentor) 'Author':'rwbL'
2021-01-25 13:26:15.675 (Garagentor) 'Name':'Garagentor'
2021-01-25 13:26:15.675 (Garagentor) 'Address':'192.168.1.225'
2021-01-25 13:26:15.675 (Garagentor) 'Port':'0'
2021-01-25 13:26:15.675 (Garagentor) 'Key':'HmIP-SWDM'
2021-01-25 13:26:15.675 (Garagentor) 'Mode1':'3597'
2021-01-25 13:26:15.675 (Garagentor) 'Mode5':'60'
2021-01-25 13:26:15.675 (Garagentor) 'Mode6':'Debug'
2021-01-25 13:26:15.675 (Garagentor) 'DomoticzVersion':'2020.2 (build 12883)'
2021-01-25 13:26:15.675 (Garagentor) 'DomoticzHash':'1a7e11b7d-modified'
2021-01-25 13:26:15.675 (Garagentor) 'DomoticzBuildTime':'2021-01-24 08:48:01'
2021-01-25 13:26:15.675 (Garagentor) Device count: 0
2021-01-25 13:26:15.675 (Garagentor) Creating new devices ...
2021-01-25 13:26:15.675 (Garagentor) Creating device 'State'.
2021-01-25 13:26:15.676 (Garagentor - State) Updating device from 0:'' to have values 1:'Closed'.
2021-01-25 13:26:15.683 (Garagentor) Device created: Garagentor - State
2021-01-25 13:26:15.683 (Garagentor) Creating new devices: OK
2021-01-25 13:26:15.683 (Garagentor) Heartbeat set: 60
2021-01-25 13:26:15.683 Python Plugin System: (Garagentor) Pushing 'PollIntervalDirective' on to queue
2021-01-25 13:26:15.683 (Garagentor) Processing 'PollIntervalDirective' message
2021-01-25 13:26:15.683 Python Plugin System: (Garagentor) Heartbeat interval set to: 60.
2021-01-25 13:26:15.386 Status: Python Plugin System: (Garagentor) Started.
2021-01-25 13:26:15.673 Status: Python Plugin System: (Garagentor) Entering work loop.
2021-01-25 13:26:15.674 Status: Python Plugin System: (Garagentor) Initialized version 1.1.1 (Build 20210125), author 'rwbL'
```

### Domoticz Device List
Idx, Hardware, Name, Type, SubType, Data
80, Garagentor, State, General, Alert, Closed

### Domoticz Log Entry with Debug=False
The plugin runs every 60 seconds (Heartbeat interval).
```
2021-01-25 13:26:25.829 (Garagentor) HMIP-SWDM: State changed to Closed
2021-01-25 13:26:25.829 (Garagentor) OPERATING_VOLTAGE=2.700000
```

```
2021-01-25 13:29:25.871 (Garagentor) HMIP-SWDM: State changed to Open
2021-01-25 13:29:25.871 (Garagentor) OPERATING_VOLTAGE=2.700000
```

## Domoticz dzVents Script alert_window-door-contact
This is an example on how to alert via email
```
--[[
    alert_window-door-contact.dzvents
    Window-Door-Contact (homematicIP device HmIP-SWDM) Alert Example.
    Send notification if the level of the Alert device changes to 4 (RED).
    20210120 rwbl
]]--

local IDX_WINDOWDOORCONTACT = 80;   -- Type General, SubType Alert

local STATE_CLOSED = 1;             -- Alert level 1 GREEN
local STATE_OPEN = 4;               -- Alert level 4 RED

return {
	on = {
		devices = {
			IDX_WINDOWDOORCONTACT
		}
	},
	execute = function(domoticz, device)
	    domoticz.log(('Device %s state changed to %s (%d).'):format(device.name, device.sValue, device.nValue), domoticz.LOG_INFO)
		if device.nValue == STATE_OPEN then
            -- Send notification via email
            local subject = ('SECURITY ALERT'):format()
            local message = ('Window-Door-Contact Garagentor %s changed to %s.'):format(device.name, device.sValue)
            domoticz.notify(subject, message, domoticz.PRIORITY_HIGH)
        end
	end
}
```

## ToDo
See TODO.md


CHANGELOG.md
# Changelog domoticz-plugin-hmip-swdm

### v1.1.1 (Build 20210125)
* UPD: Minor changes.

### v1.1.0 (Build 20210123)
* NEW: Changed datapoint attribute from ise_id to type. No need to define the datapoint ise_id anymore, only the device ise_id.

### v1.0.0 (Build 20210119)
* NEW: First version published.


plugin.py
# Domoticz Home Automation - homematicIP Window and Door Contact with Magnet (HMIP-SWDM)
# Get the state open|closed of the contact.
# Dependencies:
# RaspberryMatic XML-API CCU Addon (https://github.com/hobbyquaker/XML-API)
# Library ElementTree (https://docs.python.org/3/library/xml.etree.elementtree.html#)
# Notes:
# 1. After every change: delete the hardware using the plugin homematicIP Thermostat
# 2. After every change: restart domoticz by running from terminal the command: sudo service domoticz.sh restart
# 3. Domoticz Python Plugin Development Documentation (https://www.domoticz.com/wiki/Developing_a_Python_plugin)
# 4. Only two adevice attributes are used. The plugin is flexible to add more attributes as required (examples LOW_BAT)
#
# Author: Robert W.B. Linn
# Version: See plugin xml definition

"""
<plugin key="HmIP-SWDM" name="homematicIP Window and Door Contact with Magnet (HmIP-SWDM)" author="rwbL" version="1.1.1 (Build 20210125)">
    <description>
        <h2>homematicIP Window and Door Contact with Magnet (HmIP-SWDM) v1.1.1</h2>
        <ul style="list-style-type:square">
            <li>Get the state of the contact Open (1) or Closed (0).</li>
        </ul>
        <h3>Domoticz Devices (Type,SubType) [XML-API Device Datapoint Type]</h3>
        <ul style="list-style-type:square">
            <li>State (General,Alert) [STATE]</li>
        </ul>
        <h3>Configuration</h3>
        <ul style="list-style-type:square">
            <li>CCU IP address (default: 192.168.1.225)</li>
            <li>Device ID (default: 3597, get via XML-API script http://ccu-ip-address/addons/xmlapi/statelist.cgi)</li>
        </ul>
    </description>
    <params>
        <param field="Address" label="CCU IP" width="200px" required="true" default="192.168.1.225"/>
        <param field="Mode1" label="Device ID" width="75px" required="true" default="3597"/>
        <param field="Mode5" label="Check Interval (sec)" width="75px" required="true" default="60"/>
        <param field="Mode6" label="Debug" width="75px">
            <options>
                <option label="True" value="Debug" default="true"/>
                <option label="False" value="Normal"/>
            </options>
        </param>
    </params>
</plugin>
"""

# Set the plugin version
PLUGINVERSION = "v1.1.1"
PLUGINSHORTDESCRIPTON = "HMIP-SWDM"

## Imports
import Domoticz
import urllib
import urllib.request
from datetime import datetime
import json
import xml.etree.ElementTree as etree

## Domoticz device units used for creating & updating devices
## And the RaspberryMatic XMLAPI datapoint type
## Each of the devices have a self variable defined in function init
UNIT_STATE = 1               # TypeName: Alert
TYPE_STATE = "STATE"        

## For testing in debug mode only
TYPE_OPERATING_VOLTAGE = "OPERATING_VOLTAGE"

# User Messages - Change as required
MSGSTATECLOSED = "Closed"
MSGSTATEOPEN = "Open"

class BasePlugin:

    def __init__(self):

        # HTTP Connection
        self.httpConn = None
        self.httpConnected = 0

        # State as string as the datapoint value defined 
        self.State = "unknown"
        self.OperatingVoltage = 0
        
        # The Domoticz heartbeat is set to every 60 seconds. Do not use a higher value as Domoticz message "Error: hardware (N) thread seems to have ended unexpectedly"
        # The Soil Moisture Monitor is read every Parameter.Mode5 seconds. This is determined by using a hearbeatcounter which is triggered by:
        # (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameter.Mode5) = 0
        self.HeartbeatInterval = 60
        self.HeartbeatCounter = 0
        return

    def onStart(self):
        Domoticz.Debug(PLUGINSHORTDESCRIPTON + " " + PLUGINVERSION)
        Domoticz.Debug("onStart called")
        Domoticz.Debug("Debug Mode:" + Parameters["Mode6"])

        if Parameters["Mode6"] == "Debug":
            self.debug = True
            Domoticz.Debugging(1)
            DumpConfigToLog()

        # if there no  devices, create these
        if (len(Devices) == 0):
            Domoticz.Debug("Creating new devices ...")
            ## 1 - STATE - TypeName: Alert (Type=243, Subtype=22)
            Domoticz.Device(Name="State", Unit=UNIT_STATE, Type=243, Subtype=22, Used=1).Create()
            Devices[UNIT_STATE].Update( nValue=1, sValue=MSGSTATECLOSED )
            Domoticz.Debug("Device created: "+Devices[UNIT_STATE].Name)
            Domoticz.Debug("Creating new devices: OK")

        # Heartbeat
        Domoticz.Debug("Heartbeat set: "+Parameters["Mode5"])
        Domoticz.Heartbeat(self.HeartbeatInterval)
        return

    def onStop(self):
        Domoticz.Debug("Plugin is stopping.")

    # Send the url parameter (GET request)
    # If task = energy then to obtain device state information in xml format
    # If task = switch then switch using the self.switchstate
    # The http response is parsed in onMessage()
    def onConnect(self, Connection, Status, Description):
        Domoticz.Debug("onConnect called. Status: " + str(Status) + ", Description:" + Description)
        if (Status == 0):
            Domoticz.Debug("CCU connected successfully.")
            self.httpConnected = 1

            # request device id datapoints to get the device state
            ## url example = 'http://192.168.1.225/addons/xmlapi/state.cgi?device_id=3597'
            url = '/addons/xmlapi/state.cgi?device_id=' + Parameters["Mode1"]
                       
            Domoticz.Debug(url)
            sendData = {'Verb' : 'GET',
                        'URL' : url,
                        'Headers' : { 'Content-Type': 'text/xml; charset=utf-8', \
                                        'Connection': 'keep-alive', \
                                        'Accept': 'Content-Type: text/html; charset=UTF-8', \
                                        'Host': Parameters["Address"], \
                                        'User-Agent':'Domoticz/1.0' }
                       }
            self.httpConn.Send(sendData)
            self.httpConn.Disconnect
            return
        else:
            self.httpConnected = 0
            Domoticz.Error("HTTP connection faillure ("+str(Status)+") to: "+Parameters["Address"]+":"+Parameters["Port"]+" with error: "+Description)
            return

    # Parse the http xml response and update the domoticz devices
    def onMessage(self, Connection, Data):
        Domoticz.Debug("onMessage called")
        
        # If not conected, then leave
        if self.httpConnected == 0:
            return

        # Parse the JSON Data Object with keys Status (Number) and Data (ByteArray)
        ## 200 is OK
        responseStatus = int(Data["Status"])
        Domoticz.Debug("STATUS=responseStatus:" + str(responseStatus) + " ;Data[Status]="+Data["Status"])

        ## decode the data using the encoding as given in the xml response string
        responseData = Data["Data"].decode('ISO-8859-1')
        ## Domoticz.Debug("DATA=" + responseData)

        if (responseStatus != 200):
            Domoticz.Error("XML-API response faillure: " + str(responseStatus) + ";" + resonseData)
            return

        # Parse the xml string 
        # Get the xml tree - requires several conversions
        tree = etree.fromstring(bytes(responseData, encoding='utf-8'))
       
        # STATE
        # Get the value for the state 0 = closed, 1 = open
        # <datapoint name="HmIP-RF.001558A99D5A78:1.STATE" type="STATE" ise_id="3622" value="0" valuetype="16" valueunit="""" timestamp="1611044101"/>
        statevalue = tree.find(".//datapoint[@type='" + TYPE_STATE + "']").attrib['value']
        Domoticz.Debug(TYPE_STATE + "=" + statevalue)

        ## update the alert device if raspmatic value not equal domoticz value
        if statevalue != self.State:
            if statevalue == "0":
                Devices[UNIT_STATE].Update( nValue=1, sValue=MSGSTATECLOSED )
            if statevalue == "1":
                Devices[UNIT_STATE].Update( nValue=4, sValue=MSGSTATEOPEN )
            Domoticz.Log(PLUGINSHORTDESCRIPTON + ": State changed to " + Devices[UNIT_STATE].sValue)
            # Domoticz.Debug("S Update=" + Devices[UNIT_STATE].sValue)
        self.State = statevalue

        # OPERATING_VOLTAGE
        operatingvoltagevalue = tree.find(".//datapoint[@type='" + TYPE_OPERATING_VOLTAGE + "']").attrib['value']
        Domoticz.Log(TYPE_OPERATING_VOLTAGE + "=" + operatingvoltagevalue)
        return
        
    def onCommand(self, Unit, Command, Level, Hue):
        Domoticz.Debug("onCommand called for Unit " + str(Unit) + ": Parameter '" + str(Command) + "', Level: " + str(Level))
        return

    def onNotification(self, Name, Subject, Text, Status, Priority, Sound, ImageFile):
        Domoticz.Debug("Notification: " + Name + "," + Subject + "," + Text + "," + Status + "," + str(Priority) + "," + Sound + "," + ImageFile)

    def onDisconnect(self, Connection):
        Domoticz.Debug("onDisconnect called")

    # Check the hearbeat counter and setup the http connection to the ccu
    # After connection, there url parameter are send - see onConnect()
    def onHeartbeat(self):
        self.HeartbeatCounter = self.HeartbeatCounter + 1
        Domoticz.Debug("onHeartbeat called. Counter=" + str(self.HeartbeatCounter * self.HeartbeatInterval) + " (Heartbeat=" + Parameters["Mode5"] + ")")
        # check the heartbeatcounter against the heartbeatinterval
        if (self.HeartbeatCounter * self.HeartbeatInterval) % int(Parameters["Mode5"]) == 0:
            try:
                # Create IP connection and connect
                self.httpConn = Domoticz.Connection(Name="CCU-"+Parameters["Address"], Transport="TCP/IP", Protocol="HTTP", Address=Parameters["Address"], Port="80")
                self.httpConn.Connect()
                self.httpConnected = 0
            except:
                Domoticz.Error("IP connection failed. Check settings and restart Domoticz.")
        return
        
global _plugin
_plugin = BasePlugin()

def onStart():
    global _plugin
    _plugin.onStart()

def onStop():
    global _plugin
    _plugin.onStop()

def onConnect(Connection, Status, Description):
    global _plugin
    _plugin.onConnect(Connection, Status, Description)

def onMessage(Connection, Data):
    global _plugin
    _plugin.onMessage(Connection, Data)

def onCommand(Unit, Command, Level, Hue):
    global _plugin
    _plugin.onCommand(Unit, Command, Level, Hue)

def onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile):
    global _plugin
    _plugin.onNotification(Name, Subject, Text, Status, Priority, Sound, ImageFile)

def onDisconnect(Connection):
    global _plugin
    _plugin.onDisconnect(Connection)

def onHeartbeat():
    global _plugin
    _plugin.onHeartbeat()

#
## Generic helper functions
#

def DumpConfigToLog():
    for x in Parameters:
        if Parameters[x] != "":
            Domoticz.Debug( "'" + x + "':'" + str(Parameters[x]) + "'")
    Domoticz.Debug("Device count: " + str(len(Devices)))
    for x in Devices:
        Domoticz.Debug("Device:           " + str(x) + " - " + str(Devices[x]))
        Domoticz.Debug("Device ID:       '" + str(Devices[x].ID) + "'")
        Domoticz.Debug("Device Name:     '" + Devices[x].Name + "'")
        Domoticz.Debug("Device nValue:    " + str(Devices[x].nValue))
        Domoticz.Debug("Device sValue:   '" + Devices[x].sValue + "'")
        Domoticz.Debug("Device LastLevel: " + str(Devices[x].LastLevel))
    return

